File: src/processor.py
#!/usr/bin/env python
# encoding: utf-8

from reader import ExcelReader
from writer import ExcelWriter
from producer import ExcelProducer
import unittest
import os


class ExcelProcessor:

    def __init__(self, excelProducter):
        self.excel = excelProducter

    def Copy(self, oldRow, newRow, sheetname):
        aRow = self.excel.GetARow(oldRow, sheetname)
        self.excel.InsertRow(newRow, aRow, sheetname)

    def Save(self, filename):
        outFile = ExcelWriter(filename)
        outFile.Write(self.excel.GetBook())
        outFile.Save()


class ProcessorTest(unittest.TestCase):

    def setUp(self):
        self.excelFile = ExcelProducer()
        self.excelFile.SetBook(ExcelReader("../test/test.xls").Reads())
        self.excelProcessor = ExcelProcessor(self.excelFile)
        self.xlsFilename = "../test/tmp.xls"

    def tearDown(self):
        try:
            os.remove(self.xlsFilename)
        except OSError as e:
            print(e.message)

    def test_Copy(self):
        oriRow = 0
        newRow = 3
        sheetname = self.excelFile.GetSheetName()[0]
        self.excelProcessor.Copy(oriRow, newRow, sheetname)

        self.assertEqual(self.excelFile.GetARow(oriRow, sheetname), \
                        self.excelFile.GetARow(newRow, sheetname))

    def test_Save(self):
        sheetname = self.excelFile.GetSheetName()[0]
        self.excelProcessor.Copy(0, 3, sheetname)
        self.excelProcessor.Save(self.xlsFilename)
        self.assertEqual(self.excelFile.GetBook(), ExcelReader(self.xlsFilename).Reads())

if __name__ == '__main__':
    unittest.main()

File: src/processor.rb
#!/usr/bin/env ruby
#encoding: utf-8

require 'spreadsheet'


module Excel
  class Processor
    def initialize
      @writeBook = Spreadsheet::Workbook.new
      @default_sheet = nil
    end

    def add_sheet(sheetName)
      @default_sheet = @writeBook.create_worksheet :name => sheetName
    end

    def push_row(aRow)
      @default_sheet.insert_row(@default_sheet.row_count)
      row = @default_sheet.last_row

      aRow.length.times.each do |num|
        value = aRow[num]
        value = aRow[num].value if value.methods.include? :value
        row.push value
        row.set_format(num, aRow.format(num))
        row.height = aRow.height
      end
    end

    def copy_row_from(rb, copyRange)
      rSheet = rb.worksheet(0)

      copyRange.each do |num|
        push_row(rSheet.row(num))
      end
    end

    def Save(filename)
      @writeBook.write filename
    end
  end
end
